File: php-food-ordering/api/data_service.php
<?php

class DataService {
    private $menuItems = [];
    private $orders = [];
    private $nextOrderId = 1;
    private $dataFile = __DIR__ . '/data.json';

    public function __construct() {
        $this->loadData();
        if (empty($this->menuItems)) {
            $this->initializeMenu();
            $this->saveData();
        }
    }

    private function initializeMenu() {
        $this->menuItems = [
            [
                'id' => 1,
                'name' => 'Margherita Pizza',
                'description' => 'Classic tomato, mozzarella, and basil',
                'price' => 12.99,
                'category' => 'Pizza',
                'imageUrl' => '/images/margherita.jpg'
            ],
            [
                'id' => 2,
                'name' => 'Pepperoni Pizza',
                'description' => 'Pepperoni and mozzarella cheese',
                'price' => 14.99,
                'category' => 'Pizza',
                'imageUrl' => '/images/pepperoni.jpg'
            ],
            [
                'id' => 3,
                'name' => 'Caesar Salad',
                'description' => 'Fresh romaine lettuce with caesar dressing',
                'price' => 8.99,
                'category' => 'Salads',
                'imageUrl' => '/images/caesar.jpg'
            ],
            [
                'id' => 4,
                'name' => 'Chicken Burger',
                'description' => 'Grilled chicken breast with lettuce and mayo',
                'price' => 10.99,
                'category' => 'Burgers',
                'imageUrl' => '/images/chicken-burger.jpg'
            ],
            [
                'id' => 5,
                'name' => 'Beef Burger',
                'description' => 'Juicy beef patty with cheese and vegetables',
                'price' => 11.99,
                'category' => 'Burgers',
                'imageUrl' => '/images/beef-burger.jpg'
            ],
            [
                'id' => 6,
                'name' => 'French Fries',
                'description' => 'Crispy golden fries',
                'price' => 4.99,
                'category' => 'Sides',
                'imageUrl' => '/images/fries.jpg'
            ],
            [
                'id' => 7,
                'name' => 'Coca Cola',
                'description' => 'Refreshing cola drink',
                'price' => 2.99,
                'category' => 'Drinks',
                'imageUrl' => '/images/coke.jpg'
            ],
            [
                'id' => 8,
                'name' => 'Chocolate Cake',
                'description' => 'Rich chocolate cake slice',
                'price' => 6.99,
                'category' => 'Desserts',
                'imageUrl' => '/images/chocolate-cake.jpg'
            ]
        ];
    }

    private function loadData() {
        if (file_exists($this->dataFile)) {
            $data = json_decode(file_get_contents($this->dataFile), true);
            $this->menuItems = $data['menuItems'] ?? [];
            $this->orders = $data['orders'] ?? [];
            $this->nextOrderId = $data['nextOrderId'] ?? 1;
        }
    }

    private function saveData() {
        $data = [
            'menuItems' => $this->menuItems,
            'orders' => $this->orders,
            'nextOrderId' => $this->nextOrderId
        ];
        file_put_contents($this->dataFile, json_encode($data, JSON_PRETTY_PRINT));
    }

    public function getAllMenuItems() {
        return array_values($this->menuItems);
    }

    public function getMenuItem($id) {
        foreach ($this->menuItems as $item) {
            if ($item['id'] == $id) {
                return $item;
            }
        }
        return null;
    }

    public function getMenuItemsByCategory($category) {
        $result = [];
        foreach ($this->menuItems as $item) {
            if (strcasecmp($item['category'], $category) === 0) {
                $result[] = $item;
            }
        }
        return $result;
    }

    public function getCategories() {
        $categories = [];
        foreach ($this->menuItems as $item) {
            if (!in_array($item['category'], $categories)) {
                $categories[] = $item['category'];
            }
        }
        sort($categories);
        return $categories;
    }

    public function createOrder($cartItems, $customerName = null, $customerPhone = null, $deliveryAddress = null) {
        $order = [
            'id' => $this->nextOrderId++,
            'createdAt' => date('c'),
            'status' => 'pending',
            'items' => [],
            'total' => 0,
            'customerName' => $customerName,
            'customerPhone' => $customerPhone,
            'deliveryAddress' => $deliveryAddress
        ];

        foreach ($cartItems as $cartItem) {
            $menuItem = $this->getMenuItem($cartItem['menuItemId']);
            if ($menuItem) {
                $orderItem = [
                    'menuItemId' => $menuItem['id'],
                    'menuItemName' => $menuItem['name'],
                    'quantity' => $cartItem['quantity'],
                    'price' => $menuItem['price']
                ];
                $order['items'][] = $orderItem;
                $order['total'] += $menuItem['price'] * $cartItem['quantity'];
            }
        }

        $this->orders[] = $order;
        $this->saveData();
        return $order;
    }

    public function getAllOrders() {
        usort($this->orders, function($a, $b) {
            return strtotime($b['createdAt']) - strtotime($a['createdAt']);
        });
        return $this->orders;
    }

    public function getOrder($id) {
        foreach ($this->orders as $order) {
            if ($order['id'] == $id) {
                return $order;
            }
        }
        return null;
    }

    public function updateOrderStatus($id, $status) {
        foreach ($this->orders as &$order) {
            if ($order['id'] == $id) {
                $order['status'] = $status;
                $this->saveData();
                return $order;
            }
        }
        return null;
    }
}

?>

File: dotnet-food-ordering/Services/DataService.cs
using FoodOrdering.Models;
using System.Collections.Concurrent;

namespace FoodOrdering.Services;

public class DataService
{
    private readonly ConcurrentDictionary<int, MenuItem> _menuItems = new();
    private readonly ConcurrentDictionary<int, Order> _orders = new();
    private int _nextOrderId = 1;

    public DataService()
    {
        InitializeMenu();
    }

    private void InitializeMenu()
    {
        var menuItems = new List<MenuItem>
        {
            new() { Id = 1, Name = "Margherita Pizza", Description = "Classic tomato, mozzarella, and basil", Price = 12.99m, Category = "Pizza", ImageUrl = "/images/margherita.jpg" },
            new() { Id = 2, Name = "Pepperoni Pizza", Description = "Pepperoni and mozzarella cheese", Price = 14.99m, Category = "Pizza", ImageUrl = "/images/pepperoni.jpg" },
            new() { Id = 3, Name = "Caesar Salad", Description = "Fresh romaine lettuce with caesar dressing", Price = 8.99m, Category = "Salads", ImageUrl = "/images/caesar.jpg" },
            new() { Id = 4, Name = "Chicken Burger", Description = "Grilled chicken breast with lettuce and mayo", Price = 10.99m, Category = "Burgers", ImageUrl = "/images/chicken-burger.jpg" },
            new() { Id = 5, Name = "Beef Burger", Description = "Juicy beef patty with cheese and vegetables", Price = 11.99m, Category = "Burgers", ImageUrl = "/images/beef-burger.jpg" },
            new() { Id = 6, Name = "French Fries", Description = "Crispy golden fries", Price = 4.99m, Category = "Sides", ImageUrl = "/images/fries.jpg" },
            new() { Id = 7, Name = "Coca Cola", Description = "Refreshing cola drink", Price = 2.99m, Category = "Drinks", ImageUrl = "/images/coke.jpg" },
            new() { Id = 8, Name = "Chocolate Cake", Description = "Rich chocolate cake slice", Price = 6.99m, Category = "Desserts", ImageUrl = "/images/chocolate-cake.jpg" }
        };

        foreach (var item in menuItems)
        {
            _menuItems[item.Id] = item;
        }
    }

    public List<MenuItem> GetAllMenuItems() => _menuItems.Values.OrderBy(m => m.Id).ToList();

    public MenuItem? GetMenuItem(int id) => _menuItems.TryGetValue(id, out var item) ? item : null;

    public List<MenuItem> GetMenuItemsByCategory(string category) =>
        _menuItems.Values.Where(m => m.Category.Equals(category, StringComparison.OrdinalIgnoreCase)).ToList();

    public List<string> GetCategories() =>
        _menuItems.Values.Select(m => m.Category).Distinct().OrderBy(c => c).ToList();

    public Order CreateOrder(List<CartItem> cartItems, string? customerName, string? customerPhone, string? deliveryAddress)
    {
        var order = new Order
        {
            Id = _nextOrderId++,
            CreatedAt = DateTime.UtcNow,
            Status = "pending",
            CustomerName = customerName,
            CustomerPhone = customerPhone,
            DeliveryAddress = deliveryAddress
        };

        foreach (var cartItem in cartItems)
        {
            var menuItem = GetMenuItem(cartItem.MenuItemId);
            if (menuItem != null)
            {
                order.Items.Add(new OrderItem
                {
                    MenuItemId = menuItem.Id,
                    MenuItemName = menuItem.Name,
                    Quantity = cartItem.Quantity,
                    Price = menuItem.Price
                });
                order.Total += menuItem.Price * cartItem.Quantity;
            }
        }

        _orders[order.Id] = order;
        return order;
    }

    public List<Order> GetAllOrders() => _orders.Values.OrderByDescending(o => o.CreatedAt).ToList();

    public Order? GetOrder(int id) => _orders.TryGetValue(id, out var order) ? order : null;

    public Order? UpdateOrderStatus(int id, string status)
    {
        if (_orders.TryGetValue(id, out var order))
        {
            order.Status = status;
            return order;
        }
        return null;
    }
}

